### PR TITLE
Added a race condition test

### PR DESCRIPTION
**Pull request by Orbittman**
This is to add an extra test around the race condition bug that I had recently. I thought that it would give a good idea about how someone would implement a threaded solution. The're is nothing in the readme yet, this I will add later and the NCrunch files I'll remove. Don't bother committing this but some review comments would be welcome. There is also some refactoring of separate files that seems to have not been included, I'll refactor that at some point before we push this to master.

### CODE REVIEW

File: UnitTests/RaceConditionTests.cs [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+            foreach (var session in sessions)
+            {
+                sessionManager.Get(session.Key)
+                    .Returns(k => new Session { Key = (string)k[0], Count = sessions.Single(s => s.Key == (string)k[0]).Count });
+                sessionManager
+                    .When(x => x.Set(Arg.Any<ISession>()))
+                    .Do(s =>
+                    {
+                        Thread.Sleep(latency);
```
Comment (GrahamClark) [inline UnitTests/RaceConditionTests.cs]: should we be using `await Task.Delay(latency)` instead? Does this still cause the issue?

File: UnitTests/RaceConditionTests.cs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+                new Session { Key = Guid.NewGuid().ToString() }
+            };
+            var sessionManager = Substitute.For<ISessionManager>();
+
+            var testClass = new RaceCondition(sessionManager);
+
+            foreach (var session in sessions)
+            {
+                sessionManager.Get(session.Key)
+                    .Returns(k => new Session { Key = (string)k[0], Count = sessions.Single(s => s.Key == (string)k[0]).Count });
```
Comment (GrahamClark) [inline UnitTests/RaceConditionTests.cs]: in the NSubstitute lambdas like this, the parameter is of type `CallInfo`, which is an array of `object`s representing the parameters passed to the mocked method. So for clarity I think it's better to call these lambda parameters something like `info`, i.e. `.Returns(info => new Session...`

File: UnitTests/RaceConditionTests.cs [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+                sessionManager
+                    .When(x => x.Set(Arg.Any<ISession>()))
+                    .Do(s =>
+                    {
+                        Thread.Sleep(latency);
+                        sessions.Single(t => t.Key == ((ISession)s[0]).Key).Count = ((ISession)s[0]).Count;
```
Comment (GrahamClark) [inline UnitTests/RaceConditionTests.cs]: same here, the lambda parameters could be a bit clearer as `s` kind of implies it's a Session object or similar, when actually it's just an object array wrapper.

File: UnitTests/RaceConditionTests.cs (hunk 1 above)
Comment (Orbittman) [inline UnitTests/RaceConditionTests.cs]: It doesn't work as it's just there to simulate a pause not to change the flow of the code. Changing this breaks the simulated action.

**Comment by Orbittman** [comment]
@GrahamClark @AnturGyffrous 
I've updated the test so that all of session request run in parallel, previously they were running synchronously. The solution that we have in our code base works with this so I know the test is valid and passable. I've updated the readme to include instructions for the new test, if there are no objections I will merge this in and pass it to the .NET contractors that we are recruiting.

**Comment by AnturGyffrous** [comment]
Just a thought, when it get's merged in our comments will be visible
publicly. Would it be better to delete the pull request and branch and just
re-add the files as they are now?

**Comment by GrahamClark** [comment]
@AnturGyffrous is that because you're worried people will spot your strange use of apostrophes? :wink:

**Comment by AnturGyffrous** [comment]
Wh'ats wro'ng wit'h my a'postroph'es?

**Comment by GrahamClark** [comment]
I don't have a problem with this being public.

**Pull request merged**